nssdb/encrypt.go

```diff
@@ -0,0 +1,359 @@
+	if err != nil {
+		return nil, err
+	}
+	if len(rest) != 0 {
+		return nil, fmt.Errorf("%d bytes remaining after parsing", len(rest))
+	}
+
+	return params, nil
+}
+
+func (p pbkdf2Params) getAlgorithmIdentifier() (*algorithmIdentifier, error) {
+	b, err := asn1.Marshal(p)
+	if err != nil {
+		return nil, fmt.Errorf("marshal pbkdf2 params: %w", err)
+	}
+
+	return &algorithmIdentifier{
+		Algorithm: pbkdf2OID,
+		Parameters: asn1.RawValue{
+			FullBytes: b,
+		},
+	}, nil
+}
+
+type aes256CBCParams struct {
+	InitializationVector []byte
+}
+
+func (algID algorithmIdentifier) AsAES256CBC() (*aes256CBCParams, error) {
+	if !algID.Algorithm.Equal(aes256CBCOID) {
+		return nil, fmt.Errorf("object is %s, not %s", algID.Algorithm, aes256CBCOID)
+	}
+
+	// Parameters is the 16-byte initialization vector, not an asn1 encoded value
+	if len(algID.Parameters.FullBytes) != 16 {
+		return nil, fmt.Errorf("initialization vector is %d bytes, not 16", len(algID.Parameters.FullBytes))
+	}
+	return &aes256CBCParams{
+		InitializationVector: algID.Parameters.FullBytes,
+	}, nil
+}
+
+func (p aes256CBCParams) getAlgorithmIdentifier() (*algorithmIdentifier, error) {
+	return &algorithmIdentifier{
+		Algorithm: aes256CBCOID,
+		Parameters: asn1.RawValue{
+			FullBytes: p.InitializationVector,
+		},
+	}, nil
+}
+
+func decodeCipherText(cipherText []byte) (*encryptedDataInfo, error) {
+	edi := &encryptedDataInfo{}
+	rest, err := asn1.Unmarshal(cipherText, edi)
+	if err != nil {
+		return nil, fmt.Errorf("decode cipher text: %w", err)
+	}
+	if len(rest) > 0 {
+		return nil, fmt.Errorf("after decoding cipher text %d bytes remain", len(rest))
+	}
+
+	return edi, nil
+}
+
+// encodeCipherText takes raw encrypted data and formats it for storage along
+// with the pbes2 keygen (pbkdf2) and encryption (e.g. aes256-cbc) params used
+func encodeCipherText(encryptedData []byte, keyGenParams, encryptParams algIDGetter) ([]byte, error) {
+	keyGenAlgID, err := keyGenParams.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, fmt.Errorf("get pbes2 key gen algorithm identifier: %w", err)
+	}
+	encryptAlgID, err := encryptParams.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, fmt.Errorf("get pbes2 encryption algorithm identifier: %w", err)
+	}
+
+	pkcs5PBES2 := &pbes2Params{
+		KeyGen:     *keyGenAlgID,
+		Encryption: *encryptAlgID,
+	}
+	pbes2AlgID, err := pkcs5PBES2.getAlgorithmIdentifier()
+	if err != nil {
+		return nil, err
+	}
+
+	edi := encryptedDataInfo{
+		AlgorithmIdentifier: *pbes2AlgID,
+		EncryptedData:       encryptedData,
+	}
+	b, err := asn1.Marshal(edi)
+	if err != nil {
+		return nil, fmt.Errorf("marshal encrypted data info: %w", err)
+	}
+
+	return b, nil
+}
+
+func newAES256CBC() (*aes256CBCParams, error) {
+	iv, err := randutil.Bytes(aes.BlockSize)
+	if err != nil {
+		return nil, err
+	}
+	return &aes256CBCParams{
+		InitializationVector: iv,
+	}, nil
+}
+
+func (p aes256CBCParams) encrypt(key, plaintext []byte) ([]byte, error) {
```

Comment:
It is weird that the constructor returns pointer, but all the methods uses a value, forcing the creating of a copy of the object.

Reply:
2168975